hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmnode/RMNodeImpl.java

```diff
@@ -1160,6 +1160,11 @@ public void transition(RMNodeImpl rmNode, RMNodeEvent event) {
       // Update NM metrics during graceful decommissioning.
       rmNode.updateMetricsForGracefulDecommission(initState, finalState);
       rmNode.decommissioningTimeout = timeout;
+      // Notify NodesListManager to notify all RMApp so that each Application Master
+      // could take any required actions.
+      rmNode.context.getDispatcher().getEventHandler().handle(
+          new NodesListManagerEvent(
+              NodesListManagerEventType.NODE_USABLE, rmNode));
```

Comment:
Hmmm.. don't think we should be sending NODE_USABLE event here.. since technically, it is not usable.
Maybe consider creating a new NODE_DECOMMISSIONING event ?

Reply:
I wasn't very sure about using `NODE_USABLE`, but while I was making the changes to follow your suggestion, I have found that in the current code [`TestRMNodeTransitions.testResourceUpdateOnDecommissioningNode`](https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRMNodeTransitions.java#L994) is asserting that `NodesListManagerEventType.NODE_USABLE` is expected for a node that transitions to `DECOMMISSIONING`. Also, NodesListManagerEventType is transformed into the corresponding `RMAppNodeUpdateType` in `NodesListManager.handle` to build a `RMAppNodeUpdateEvent` that is processed in `RMAppImpl.processNodeUpdate` which just uses the `RMAppNodeUpdateType` for logging.

So it looks like it is ok to use `NodesListManagerEventType.NODE_USABLE` for nodes in decommissioning state. Do you still think it's worth adding some additional value for `NodesListManagerEventType` and `RMAppNodeUpdateType`?

Reply:
I feel we should make intentions explicit - having a separate event type might make the code cleaner and easier to follow - rather than overloading. It could be that the assumption in the testcase is wrong (will have to double check though), in which case - it is perfectly alright to modify the testcasse with the new event.

Reply:
Makes sense, I have added a new commit for adding a new value `NODE_DECOMMISSIONING` for `NodesListManagerEventType`

Reply:
Thanks - looking at the patch.. can you also attach a consolidated patch on the JIRA ? So as to kick Jenkins.

Reply:
I have just attached the patch. Thanks a lot for taking a look!